core/model/modx/processors/resource/trash/restore.class.php

```diff
@@ -0,0 +1,196 @@
+<?php
+
+/**
+ * Restores deleted files.
+ *
+ * @return boolean
+ * @package    modx
+ * @subpackage processors.resource
+ */
+class modResourceTrashRestoreProcessor extends modProcessor {
+
+    /** @var modResource $resource */
+    public $resource;
+
+    private $resourceList;
+    private $idList;
+    private $contexts;
+
+    /** @var array Failed ids of restored resources */
+    private $failures;
+
+    /** @var array Ids of successfully restored resources */
+    private $success;
+
+    public function checkPermissions() {
+        return $this->modx->hasPermission('undelete_document');
+    }
+
+    public function getLanguageTopics() {
+        return array('resource', 'trash');
+    }
+
+    /**
+     * @return bool|null|string
+     */
+    public function initialize() {
+        // we expect a list of ids here
+        $idlist = $this->getProperty('ids', false);
+
+        $this->contexts = array();
+
+        if (!$idlist) {
+            return $this->modx->lexicon('resource_err_ns');
+        }
+
+        $this->idList = explode(',', $idlist);
+
+        if (count($this->idList) === 0) {
+            return $this->modx->lexicon('resource_err_ns');
+        }
+
+        // and now retrieve a collection of resources here
+        $this->resourceList = $this->modx->getCollection('modResource', array(
+            'deleted' => true,
+            'id:IN' => $this->idList,
+        ));
+        $count = count($this->resourceList);
+
+        if ($count === 0) {
+            return $this->modx->lexicon('resource_err_nfs', array('ids' => $this->idList));
+        }
+
+        return true;
+    }
+
+    public function process() {
```

Comment:
I suggest you to replace it to something like this:
```
foreach ($ids as $id) {
    $response = $this->modx->runProcessor('resource/undelete', array('id' => $id));
    if ($response->isError()) {
        return $response->getResponse();;
    }
}
return $this->modx->success();
```
We already have undelete processor logic, why just don`t use it? It is much simpler and allows to avoid duplicate of logic.

And maybe it is time to add `resource/remove` processor and use it in `resource/emptyrecyclebin` and `resource/trash/purge`